DigitalLearningSolutions.Web/Helpers/RegistrationEmailValidator.cs

```diff
@@ -0,0 +1,104 @@
+            IUserService userService,
+            ICentresDataService centresDataService
+        )
+        {
+            ValidateEmailAddresses(true, model, modelState, userService, centresDataService);
+        }
+
+        private static void ValidateEmailAddresses(
+            bool isRegisterAdminJourney,
+            PersonalInformationViewModel model,
+            ModelStateDictionary modelState,
+            IUserService userService,
+            ICentresDataService? centresDataService = null
+        )
+        {
+            var primaryEmailIsValidAndNotNull =
+                !modelState.HasError(nameof(PersonalInformationViewModel.PrimaryEmail)) &&
+                model.PrimaryEmail != null;
+            var centreSpecificEmailIsValidAndNotNull =
+                !modelState.HasError(nameof(PersonalInformationViewModel.CentreSpecificEmail)) &&
+                model.CentreSpecificEmail != null;
+
+            var autoRegisterManagerEmail = isRegisterAdminJourney
+                ? centresDataService!.GetCentreAutoRegisterValues(model.Centre!.Value)
+                    .autoRegisterManagerEmail
+                : null;
+
+            if (primaryEmailIsValidAndNotNull)
+            {
+                if (userService.EmailIsInUse(model.PrimaryEmail))
+                {
+                    modelState.AddModelError(
+                        nameof(PersonalInformationViewModel.PrimaryEmail),
+                        DuplicateEmailErrorMessage
+                    );
+                }
+                else if (isRegisterAdminJourney)
+                {
+                    if (EmailsDoNotMatch(model.PrimaryEmail, autoRegisterManagerEmail!) && model.CentreSpecificEmail == null)
+                    {
+                        modelState.AddModelError(
+                            nameof(PersonalInformationViewModel.PrimaryEmail),
+                            WrongEmailForCentreErrorMessage
+                        );
+                    }
+                }
+            }
+
+            if (centreSpecificEmailIsValidAndNotNull)
+            {
+                if (userService.EmailIsInUse(model.CentreSpecificEmail))
+                {
+                    modelState.AddModelError(
+                        nameof(PersonalInformationViewModel.CentreSpecificEmail),
+                        DuplicateEmailErrorMessage
+                    );
+                }
+                else if (isRegisterAdminJourney && EmailsDoNotMatch(model.PrimaryEmail!, autoRegisterManagerEmail!) &&
+                         EmailsDoNotMatch(model.CentreSpecificEmail, autoRegisterManagerEmail))
+                {
+                    modelState.AddModelError(
+                        nameof(PersonalInformationViewModel.CentreSpecificEmail),
+                        WrongEmailForCentreErrorMessage
+                    );
+                }
+            }
+        }
+
+        private static bool EmailsDoNotMatch(string email, string autoRegisterManagerEmail)

```

Comment:
Some improvements that could be made here to make the method and its calling code more readable and generic:
* Negate it to void double negatives in future usages
* Call it `StringsMatchCaseInsensitive` (the implementation doesn't care if the strings are emails)
* Make the parameters nullable (string.Equals accepts nullable strings)
* Pull it out into a helper so it can be used elsewhere

Reply:
Sounds good, done!